compiler_opt/rl/imitation_learning/weighted_bc_trainer.py

```diff
@@ -0,0 +1,535 @@
+from compiler_opt.rl.inlining import imitation_learning_config as config
+from compiler_opt.rl import feature_ops
+
+import keras
+import numpy as np
+import tensorflow as tf
+
+import tf_agents
+from tf_agents.trajectories import time_step as ts
+from tf_agents.typing import types
+from tf_agents.trajectories import policy_step
+import tensorflow_probability as tfp
+
+_QUANTILE_MAP_PATH = flags.DEFINE_string(
+    'quantile_map_path', None,
+    ('Directory containing the quantile map for normalizing features'
+     'in feature_ops.build_quantile_map.'))
+_TRAINING_DATA = flags.DEFINE_multi_string(
+    'training_data', None, 'Training data for one step of BC-Max')
+_PROFILING_DATA = flags.DEFINE_multi_string(
+    'profiling_data', None,
+    ('Paths to profile files for computing the TrainingWeights'
+     'If specified the order for each pair of json files is'
+     'comparator.json followed by eval.json and the number of'
+     'files should always be even.'))
+_SAVE_MODEL_DIR = flags.DEFINE_string(
+    'save_model_dir', None, 'Location to save the keras and TFAgents policies.')
+# _GIN_FILES = flags.DEFINE_multi_string(
+#     'gin_files', [], 'List of paths to gin configuration files.')
+# _GIN_BINDINGS = flags.DEFINE_multi_string(
+#     'gin_bindings', [],
+#     'Gin bindings to override the values set in the config files.')
+
+FLAGS = flags.FLAGS
+
+# Pytype cannot pick up the pyi file for tensorflow.summary. Disable the error
+# here as these errors are false positives.
+# pytype: disable=pyi-error
+
+
+@gin.configurable
+class TrainingWeights:
+  """Class for computing weights for training."""
+
+  def __init__(
+      #  pylint: disable=dangerous-default-value
+      self,
+      partitions: list[float] = [0.],
+      weights: Optional[np.ndarray] = None):
+    self._weights = weights
+    if not weights:
+      self._weights = np.ones(len(partitions) + 1)
+    self._probs: np.ndarray = np.exp(self._weights) / np.sum(
+        np.exp(self._weights))
+    self._partitions: List[float] = partitions
+    self._round: int = 1
+
+  def _bucket_by_feature(
+      self, data: List[ProfilingDictValueType],
+      feature_name: str) -> List[List[ProfilingDictValueType]]:
+    """Partitions the profiles according to the feature name.
+    
+    Partitions the profiles according to the feature name and the
+    buckets defined by self._partitions.
+
+    Args:
+      data: list of ProfilingDictValueType to partition
+        feature_name: feature according to which the partition happens
+
+    Returns:
+      buckets: partitioned profiles according to the feature name
+    """
+    buckets = [[] for i in range(len(self._partitions) + 1)]
+
+    for prof in data:
+      idx = bisect.bisect_right(self._partitions, prof[feature_name])
+      buckets[idx].append(prof)
+
+    return buckets
+
+  def _eg_step(self, loss, step_size) -> np.ndarray:
```

Comment:
can you rename `_eg_step` to something more verbose for clarity? `_exp_gradient_step` for instance?

Reply:
also, function names should be verbs unless they are trivial accessors. So `_get_exp_gradient_step`?